File: app/src/main/java/com/example/project2/db/pokemon/PokemonDao.java
package com.example.project2.db.pokemon;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import java.util.List;

@Dao
public interface PokemonDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Pokemon... pokemon);

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    long insert(Type type);

    @Query("SELECT * FROM `type` WHERE name = :name")
    Type getTypeByName(String name);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(PokemonType... pokemonTypes);

    @Query("SELECT * FROM pokemon WHERE pokemon_id = :pokemonId")
    Pokemon getPokemonById(int pokemonId);

    @Query("SELECT * FROM pokemon ORDER BY RANDOM() LIMIT 1")
    Pokemon getRandomPokemon();

    @Query("SELECT T.name FROM `type` T JOIN pokemon_type PT ON T.type_id = PT.type_id WHERE PT.pokemon_id = :pokemonId")
    List<String> getTypesForPokemon(int pokemonId);
}

File: app/src/main/java/com/example/project2/data/dao/PokemonDao.kt

package com.example.project2.data.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Transaction
import com.example.project2.data.Pokemon
import com.example.project2.data.PokemonWithMoves
import com.example.project2.data.PokemonWithTypes

@Dao
interface PokemonDao {
    @Insert
    suspend fun insert(pokemon: Pokemon)

    @Transaction
    @Query("SELECT * FROM pokemon")
    fun getPokemonWithTypes(): List<PokemonWithTypes>

    @Transaction
    @Query("SELECT * FROM pokemon")
    fun getPokemonWithMoves(): List<PokemonWithMoves>
}
